Implement at() in terms of operator[] to avoid bound checks in deque and vector

The containers in the STL that support random access implement two different functions to access specific elements in the container:

- `operator[]`
- `at`

`operator[]` just accesses the element without bound checks, while `at` perfoms the same operation, but doing bound checks and throwing an out_of_bounds exception when the index is out of bounds.

I propose that the containers (deque and vector basically) implement `at()` in terms of `operator[]`, and perfom bound checks only when calling `at()`. That way they're more compliant to the STL ones. This is my proposed solution for vector (for deque is the same idea):

```
template <typename T>
inline STDGPU_DEVICE_ONLY typename vector<T>::reference
vector<T>::at(const vector<T>::index_type n)
{
    return const_cast<vector<T>::reference>(static_cast<const vector<T>*>(this)->at(n));
}


template <typename T>
inline STDGPU_DEVICE_ONLY typename vector<T>::const_reference
vector<T>::at(const vector<T>::index_type n) const
{
    STDGPU_EXPECTS(0 <= n);
    STDGPU_EXPECTS(n < size());
    STDGPU_EXPECTS(occupied(n));

    return this->operator[](n);
}


template <typename T>
inline STDGPU_DEVICE_ONLY typename vector<T>::reference
vector<T>::operator[](const vector<T>::index_type n)
{
    return _data[n];
}


template <typename T>
inline STDGPU_DEVICE_ONLY typename vector<T>::const_reference
vector<T>::operator[](const vector<T>::index_type n) const
{
    return _data[n];
}
```

CI breakage is actually unrelated to the changes as installing the required packages already fails:

https://dev.azure.com/patrickstotko/stdgpu/_build/results?buildId=409&view=logs&j=26ef8efa-a13f-5927-0bd6-f52d59ca1b8a&t=6e8fff25-1341-5e4b-5d44-a0c41a8a32ee&l=74

```
Reading package lists...
E: Failed to fetch http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_18.04/InRelease  403  Forbidden [IP: 195.135.221.134 80]
E: The repository 'http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_18.04  InRelease' is no longer signed.
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://storage.googleapis.com/bazel-apt stable InRelease: The following signatures were invalid: EXPKEYSIG 3D5919B448457EE0 Bazel Developer (Bazel APT repository key) <bazel-dev@googlegroups.com>

##[error]Bash exited with code '100'.
```

We need to wait until the CI images get either reverted or fixed.

--- a/src/stdgpu/impl/deque_detail.cuh
+++ b/src/stdgpu/impl/deque_detail.cuh
@@ -93,11 +93,7 @@ deque<T>::at(const deque<T>::index_type n) const
     STDGPU_EXPECTS(n < size());
     STDGPU_EXPECTS(occupied(n));
 
-    index_t index_to_wrap = static_cast<index_t>(_begin.load()) + n;
-
-    STDGPU_ASSERT(0 <= index_to_wrap);
-
-    return _data[index_to_wrap % _capacity];
+    return operator[](n);
 }
 
 
@@ -105,7 +101,7 @@ template <typename T>
 inline STDGPU_DEVICE_ONLY typename deque<T>::reference
 deque<T>::operator[](const deque<T>::index_type n)
 {
-    return at(n);
+    return const_cast<reference>(static_cast<const deque<T>*>(this)->operator[](n));
 }
 
 
@@ -113,7 +109,8 @@ template <typename T>
 inline STDGPU_DEVICE_ONLY typename deque<T>::const_reference
 deque<T>::operator[](const deque<T>::index_type n) const
 {
-    return at(n);
+    index_t index_to_wrap = static_cast<index_t>(_begin.load()) + n;
+    return _data[index_to_wrap % _capacity];
 }
 
 

--- a/src/stdgpu/impl/vector_detail.cuh
+++ b/src/stdgpu/impl/vector_detail.cuh
@@ -82,7 +82,7 @@ vector<T>::at(const vector<T>::index_type n) const
     STDGPU_EXPECTS(n < size());
     STDGPU_EXPECTS(occupied(n));
 
-    return _data[n];
+    return operator[](n);
 }
 
 
@@ -90,7 +90,7 @@ template <typename T>
 inline STDGPU_DEVICE_ONLY typename vector<T>::reference
 vector<T>::operator[](const vector<T>::index_type n)
 {
-    return at(n);
+    return const_cast<vector<T>::reference>(static_cast<const vector<T>*>(this)->operator[](n));
 }
 
 
@@ -98,7 +98,7 @@ template <typename T>
 inline STDGPU_DEVICE_ONLY typename vector<T>::const_reference
 vector<T>::operator[](const vector<T>::index_type n) const
 {
-    return at(n);
+    return _data[n];
 }
 
 

